### PR TITLE
Replace manual memoization by functools.lru_cache.

### PR DESCRIPTION
**Pull request by anntzer**

### CODE REVIEW

File: defopt.py [hunk 1]
```diff
@@ -25,12 +25,14 @@
 from sphinxcontrib.napoleon.docstring import GoogleDocstring, NumpyDocstring
 import typing_inspect as ti
 
-try:
-    import collections.abc as collections_abc
-    from inspect import Parameter, Signature, signature as _inspect_signature
-except ImportError:  # pragma: no cover
+if sys.version_info < (3,):  # pragma: no cover
```
Comment (evanunderscore) [inline defopt.py]: I'm usually more inclined to put the Python 3 code before the Python 2 code. Also personally think this is clearer as if sys.version_info.major > 3.

File: setup.py [hunk 2]
```diff
@@ -23,6 +23,7 @@
         'typing_inspect>=0.3.1',
     ],
     extras_require={
+        ':python_version<"3.2"': ['backports.functools_lru_cache'],
```
Comment (evanunderscore) [inline setup.py]: Not sure if this change is really worth adding another install dependency for 2.7. I'm not planning to support any EOL Python versions past 2020-01-01, which means we can drop anything below 3.5. What do you think?

Comment (anntzer) [reply setup.py]: I have stopped using py2 years ago, so I basically don't care about it.  We can just delay this PR until 2020-01-01, or drop py2 support now, or put in this PR as is for now, whatever you prefer.

File: defopt.py (hunk 1 above)
Comment (anntzer) [reply defopt.py]: inverted it

File: setup.py (hunk 2 above)
Comment (evanunderscore) [reply setup.py]: I'd rather hold off on this for the moment. Once I've tagged and published 5.2, I'd be happy to start making backwards-incompatible changes on master in preparation for 6.0. If necessary, we can branch off 5.2 to make bug fixes on request.

Comment (anntzer) [reply setup.py]: OK, let's just leave this as is for now.

**Comment by anntzer** [comment]
Let's close this and have a single PR to strip py<3 compat when the time comes.

**Pull request closed without merging**